File: rust/src/figures.rs
//! Figure image extraction and download module (R4).
//!
//! Extracts figure images from web pages and downloads them locally.
//! Supports multi-language figure detection (English/Russian).
//!
//! Based on reference implementation from:
//! <https://github.com/link-foundation/meta-theory/blob/main/scripts/download.mjs>

use regex::Regex;
use scraper::{Html, Selector};
use serde::{Deserialize, Serialize};
use url::Url;

/// A figure extracted from HTML.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Figure {
    pub figure_num: u32,
    pub src: String,
    pub alt: String,
    pub caption: String,
    pub sequential_index: u32,
}

/// Result of downloading a single figure.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FigureDownloadResult {
    pub figure_num: u32,
    pub filename: String,
    #[serde(skip)]
    pub buffer: Option<Vec<u8>>,
    pub caption: String,
    pub original_url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Extract figure elements from HTML content.
///
/// Finds `<figure>` elements containing `<img>` tags and extracts
/// their source URLs, alt text, and captions with multi-language
/// figure number detection.
#[must_use]
pub fn extract_figures(html: &str, base_url: &str) -> Vec<Figure> {
    let document = Html::parse_document(html);
    let mut figures = Vec::new();
    let mut sequential_index: u32 = 0;

    let Ok(figure_sel) = Selector::parse("figure") else {
        return figures;
    };
    let img_sel = Selector::parse("img").unwrap();
    let caption_sel = Selector::parse("figcaption").unwrap();

    let figure_num_re = Regex::new(r"(?i)(?:Figure|Рис\.?|Рисунок)\s*(\d+)").unwrap();

    for figure_el in document.select(&figure_sel) {
        let Some(img) = figure_el.select(&img_sel).next() else {
            continue;
        };

        let src = match img.value().attr("src") {
            Some(s) if !s.starts_with("data:") && !s.contains(".svg") => s,
            _ => continue,
        };

        sequential_index += 1;

        let caption_text = figure_el
            .select(&caption_sel)
            .next()
            .map(|el| el.text().collect::<String>().trim().to_string())
            .unwrap_or_default();

        let figure_num = figure_num_re
            .captures(&caption_text)
            .and_then(|cap| cap[1].parse::<u32>().ok())
            .unwrap_or(sequential_index);

        let resolved_src = Url::parse(base_url)
            .ok()
            .and_then(|base| base.join(src).ok())
            .map_or_else(|| src.to_string(), |u| u.to_string());

        let alt = img.value().attr("alt").unwrap_or("").to_string();

        figures.push(Figure {
            figure_num,
            src: resolved_src,
            alt,
            caption: caption_text,
            sequential_index,
        });
    }

    figures
}

/// Download figure images.
///
/// Downloads each figure's image and returns results with the buffer
/// or error information.
pub async fn download_figures(figures: &[Figure]) -> Vec<FigureDownloadResult> {
    let mut results = Vec::new();
    let client = reqwest::Client::new();

    for figure in figures {
        let ext = if figure.src.contains(".jpeg") || figure.src.contains(".jpg") {
            "jpg"
        } else {
            "png"
        };
        let filename = format!("figure-{}.{ext}", figure.figure_num);

        let mut last_error = None;
        let mut buffer = None;

        for attempt in 0..3 {
            match client.get(&figure.src).send().await {
                Ok(resp) if resp.status().is_success() => match resp.bytes().await {
                    Ok(bytes) => {
                        buffer = Some(bytes.to_vec());
                        break;
                    }
                    Err(e) => last_error = Some(e.to_string()),
                },
                Ok(resp) => last_error = Some(format!("HTTP {}", resp.status())),
                Err(e) => last_error = Some(e.to_string()),
            }
            if attempt < 2 {
                tokio::time::sleep(std::time::Duration::from_secs(1)).await;
            }
        }

        let has_buffer = buffer.is_some();
        results.push(FigureDownloadResult {
            figure_num: figure.figure_num,
            filename,
            buffer,
            caption: figure.caption.clone(),
            original_url: figure.src.clone(),
            error: if has_buffer { None } else { last_error },
        });
    }

    results
}

File: js/src/figures.js
/**
 * Figure image extraction and download module (R4).
 *
 * Extracts figure images from web pages and downloads them locally.
 * Supports multi-language figure detection (English/Russian).
 *
 * Based on reference implementation from:
 * https://github.com/link-foundation/meta-theory/blob/main/scripts/download.mjs
 *
 * @module figures
 */

import fetch from 'node-fetch';
import * as cheerio from 'cheerio';
import { URL } from 'url';
import { retry } from './retry.js';
import { createBrowser, getBrowserEngine } from './browser.js';
import { dismissPopups, scrollToLoadContent } from './popups.js';

/**
 * Extract figure elements from HTML content.
 *
 * @param {string} html - HTML content
 * @param {string} baseUrl - Base URL for resolving relative URLs
 * @returns {Object[]} Array of figure objects with src, alt, caption, figureNum
 */
export function extractFigures(html, baseUrl) {
  const $ = cheerio.load(html);
  const figures = [];
  let sequentialIndex = 0;

  $('figure').each(function () {
    const img = $(this).find('img').first();
    if (!img.length) {
      return;
    }

    const src = img.attr('src');
    if (!src || src.startsWith('data:') || src.includes('.svg')) {
      return;
    }

    sequentialIndex++;
    const captionEl = $(this).find('figcaption');
    const captionText = captionEl.length ? captionEl.text().trim() : '';

    // Match multi-language figure numbers: "Figure X", "Рис. X", "Рисунок X"
    const figureMatch = captionText.match(/(?:Figure|Рис\.?|Рисунок)\s*(\d+)/i);
    const figureNum = figureMatch ? parseInt(figureMatch[1]) : sequentialIndex;

    let resolvedSrc;
    try {
      resolvedSrc = new URL(src, baseUrl).href;
    } catch {
      resolvedSrc = src;
    }

    figures.push({
      figureNum,
      src: resolvedSrc,
      alt: img.attr('alt') || '',
      caption: captionText,
      sequentialIndex,
    });
  });

  return figures;
}

/**
 * Download figure images to a local directory.
 *
 * @param {Object[]} figures - Array of figure objects from extractFigures
 * @param {Object} [options] - Download options
 * @param {Function} [options.onProgress] - Callback(figureNum, status)
 * @returns {Promise<Object[]>} Array of download results with metadata
 */
export async function downloadFigures(figures, options = {}) {
  const results = [];

  for (const figure of figures) {
    const ext =
      figure.src.includes('.jpeg') || figure.src.includes('.jpg')
        ? 'jpg'
        : 'png';
    const filename = `figure-${figure.figureNum}.${ext}`;

    try {
      const resp = await retry(() => fetch(figure.src), {
        retries: 3,
        baseDelay: 1000,
      });

      if (!resp.ok) {
        throw new Error(`HTTP ${resp.status}`);
      }

      const buffer = await resp.buffer();

      if (options.onProgress) {
        options.onProgress(figure.figureNum, 'downloaded');
      }

      results.push({
        figureNum: figure.figureNum,
        filename,
        buffer,
        caption: figure.caption,
        originalUrl: figure.src,
      });
    } catch (err) {
      if (options.onProgress) {
        options.onProgress(figure.figureNum, 'failed');
      }
      results.push({
        figureNum: figure.figureNum,
        filename,
        buffer: null,
        caption: figure.caption,
        originalUrl: figure.src,
        error: err.message,
      });
    }
  }

  return results;
}

/**
 * Extract and download figures from a URL using browser rendering.
 * This handles JavaScript-rendered pages that require a browser.
 *
 * @param {string} url - URL to extract figures from
 * @param {Object} [options] - Options
 * @param {string} [options.engine] - Browser engine
 * @returns {Promise<Object>} Result with figures and metadata
 */
export async function extractFiguresFromUrl(url, options = {}) {
  const absoluteUrl = url.startsWith('http') ? url : `https://${url}`;
  const engine = options.engine || 'puppeteer';

  const browser = await createBrowser(engine);
  try {
    const page = await browser.newPage();
    await page.setUserAgent(
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    );
    await page.setViewport({ width: 1920, height: 1080 });
    await page.goto(absoluteUrl, {
      waitUntil: 'networkidle0',
      timeout: 30000,
    });

    await new Promise((resolve) => setTimeout(resolve, 5000));
    await scrollToLoadContent(page);
    await dismissPopups(page);

    const html = await page.content();
    const figures = extractFigures(html, absoluteUrl);
    const downloaded = await downloadFigures(figures, options);

    return {
      url: absoluteUrl,
      figures: downloaded,
      totalFound: figures.length,
      totalDownloaded: downloaded.filter((d) => d.buffer).length,
    };
  } finally {
    await browser.close();
  }
}

/**
 * Figures extraction handler for Express API.
 *
 * Query parameters:
 *   url    (required) - URL to extract figures from
 *   engine - 'puppeteer' or 'playwright'
 */
export async function figuresHandler(req, res) {
  const url = req.query.url;
  if (!url) {
    return res.status(400).send('Missing `url` parameter');
  }

  try {
    const engine = getBrowserEngine(req);
    const result = await extractFiguresFromUrl(url, { engine });

    // Return metadata (not the actual image buffers via JSON)
    res.json({
      url: result.url,
      totalFound: result.totalFound,
      totalDownloaded: result.totalDownloaded,
      figures: result.figures.map((f) => ({
        figureNum: f.figureNum,
        filename: f.filename,
        caption: f.caption,
        originalUrl: f.originalUrl,
        downloaded: !!f.buffer,
        error: f.error || null,
      })),
    });
  } catch (err) {
    console.error(err);
    res.status(500).send('Error extracting figures');
  }
}
